crates/recorder/src/benchmark.rs

```diff
@@ -119,68 +124,100 @@ impl<'a, 'b> Instance<'a, 'b> {
     // as imports, we must retain the measurement state here, in the host code--see `static_measure`.
-    let execution = instance.execute(measure);
-    info!("Executed successfully (bench time): {:?}", execution);
-
-    Ok(BenchmarkMeasurements {
-        compilation,
-        instantiation,
-        execution,
-    })
-}
+    instance.execute(measure, measurements);
+    info!("Executed successfully");
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct BenchmarkMeasurements {
-    compilation: Measurement,
-    instantiation: Measurement,
-    execution: Measurement,
+    Ok(())
 }
 
-// Because this recorder cedes control to the Wasm module during execution, giving it the responsibility for calling
-// `bench_start` and `bench_end` at the appropriate times, we need to maintain any recording state here. This Rust
-// module does so using static variables, naturally resulting in unsafe behavior. This module should be safe if used
-// correctly:
-//  - `setup()` the static values
-//  - call `bench_start()` and then `bench_end()`
-//  - retrieve the measurement with `result()`
+/// Because this recorder cedes control to the Wasm module during execution,
+/// giving it the responsibility for calling `bench_start` and `bench_end` at
+/// the appropriate times, we need to maintain any recording state here. This
+/// Rust module does so using atomic pointers and internal unsafe code. A safe
+/// interface is provided via the `with` function.
 mod static_measure {
     use super::*;
-    static mut MEASURE: Option<Box<dyn Measure>> = None;
-    static mut MEASUREMENT: Option<Measurement> = None;
-    pub(crate) fn setup(measure: Box<dyn Measure>) {
-        unsafe {
-            MEASURE = Some(measure);
-            MEASUREMENT = None;
+    use std::ptr;
+    use std::sync::atomic::{AtomicPtr, Ordering};
+
+    static STATE: AtomicPtr<MeasureState> = AtomicPtr::new(ptr::null_mut());
+
+    /// Call `f` with the static measurement state used by `bench_{start,end}`
+    /// initialized to the given measure and measurements pair.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the static measurment state is already initialized, which
+    /// could only happen if this function were called on multiple threads
+    /// concurrently, which is not allowed.
+    pub(crate) fn with<T>(
+        measure: &mut dyn Measure,
+        measurements: &mut Measurements,
+        f: impl FnOnce() -> T + std::panic::UnwindSafe,
+    ) -> T {
+        let measure: *mut dyn Measure = measure as _;
+        let measure = ptr::NonNull::new(measure).unwrap();
+        let measurements = ptr::NonNull::new(measurements).unwrap();
+        let measurements = measurements.cast();
+
+        let state = &mut MeasureState {
+            measure,
+            measurements,
+        };
+
+        let old_state = STATE.swap(state, Ordering::SeqCst);
+        assert!(old_state.is_null());
+
+        let result = std::panic::catch_unwind(f);
```

Comment:
Learn something new about Rust every day... These are good changes! I would prefer if all of this code did not know about `Measurements`, but the concurrency handling here is great.